File: data-structures/stacks/c-sharp/stack_linked_list_OOP.cs
/*
Raspberry Pi Foundation
Developed to be used alongside Isaac Computer Science, part of the National Centre for Computing Education
Usage licensed under CC BY-SA 4

Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;

namespace IsaacCodeSamples
{
    // A class for a node in a linked list
    class Node
    {
        private string data;
        private Node next;

        // Constructor method
        public Node(string itemData)
        {
            data = itemData;
        }

        public string GetData()
        {
            return data;
        }

        public Node GetNext()
        {
            return next;
        }

        public void SetNext(Node newNext)
        {
            next = newNext;
        }

        // Testing
        public void OutputNode()
        {
            string nextData = "null";
            if (next != null)
                nextData = next.data;
            Console.WriteLine($"Data: {data}, next: {nextData}");
        }
    }


    // A stack implemented as a linked list
    class Stack
    {
        private Node top;

        // Check if the stack is empty
        public bool IsEmpty()
        {
            if (top == null)
                return true;
            else
                return false;
        }

        // Push a new item onto the stack
        public void Push(string data)
        {
            Node newNode = new Node(data);

            // Check if the stack is NOT empty
            if (!IsEmpty()) {
                // Point to the next element in the list
                newNode.SetNext(top);
            }
            
            // Set top to point to the new node
            top = newNode;
        }

        // Pop the item from the top of the stack
        public string Pop()
        {
            string poppedItem;

            if (IsEmpty()) {
                Console.WriteLine("\nStack is empty - nothing to pop");
                poppedItem = "";
            }
            else {
                // Get data and change pointer to next item
                poppedItem = top.GetData();
                top = top.GetNext();
            }

            return poppedItem;
        }

        // Peek the item at the top of the stack
        public string Peek()
        {
            string peekedItem;

            if (IsEmpty()) {
                Console.WriteLine("\nStack is empty - nothing to peek");
                peekedItem = "";
            }
            else {
                // Get data from the top of the stack
                peekedItem = top.GetData();
            }

            return peekedItem;
        }

        // Output the state of the stack
        public void OutputStack()
        {
            // Testing
            if (top != null) {
                Console.WriteLine("------ State of the stack ------");
                Node currentNode = top;
                currentNode.OutputNode();
                while (currentNode.GetNext() != null) {
                    currentNode = currentNode.GetNext();
                    currentNode.OutputNode();
                }
                Console.WriteLine();
            }
        }
    }


    class StacksExample
    {
        // The Main method is the entry point for all C# programs
        public static void Main()
        {
            Console.WriteLine("### Stack (OOP) ###\n");

            // Initialise the stack
            Stack myStack = new Stack();

            // Insert test data onto the stack
            myStack.Push("Julie");
            myStack.OutputStack();
            
            myStack.Push("Rey");
            myStack.OutputStack();

            // Peek the top item of the stack
            string peekedItem = myStack.Peek();
            Console.WriteLine($"Peeked top of the stack: {peekedItem}");
            myStack.OutputStack();
            
            // Insert more test data onto the stack
            myStack.Push("Habib");
            myStack.OutputStack();
            
            myStack.Push("Sabrina");
            myStack.OutputStack();

            // Pop the top item from the stack
            string poppedItem = myStack.Pop();
            Console.WriteLine($"Popped {poppedItem}");
            myStack.OutputStack();
            
            // Insert more test data onto the stack
            myStack.Push("Mina");
            myStack.OutputStack();
            
            myStack.Push("Eirini");
            myStack.OutputStack();
        }

    }
}

File: data-structures/stacks/python/stack_linked_list_OOP.py
# Raspberry Pi Foundation
# Developed to be used alongside Isaac Computer Science,
# part of the National Centre for Computing Education
# Usage licensed under CC BY-SA 4


class Node():
    """A node in a linked list"""

    # Constructor method
    def __init__(self, data):
        self.data = data
        self.next = None

    def get_data(self):
        return self.data

    def get_next(self):
        return self.next

    def set_next(self, new_next):
        self.next = new_next
        
    def output_node(self):
        """Testing"""
        next_data = None
        if (self.next != None):
            next_data = self.next.data
        print(f"Data: {self.data}, next: {next_data}") 


class Stack():
    """A stack implemented as a linked list"""
    
    # Constructor method
    def __init__(self):
        self.top = None

    
    def is_empty(self):
        """Check if the stack is empty"""
        if self.top == None:
            return True
        else:
            return False


    def push(self, data):
        """Push a new item onto the stack"""
        
        new_node = Node(data)
        
        # Check if the stack is NOT empty
        if not self.is_empty():
             # Point to the next element in the list
            new_node.set_next(self.top)

        # Set top to point to the new node
        self.top = new_node         


    def pop(self):
        """Pop the item from the top of the stack"""
        
        popped_data = None
        
        if self.is_empty():
            print("\nStack is empty - nothing to pop")
        else:
             # Get data and change pointer to next item
            popped_data = self.top.get_data()
            self.top = self.top.get_next()
            
        return popped_data


    def peek(self):
        """Peek the item at the top of the stack"""
        
        peeked_data = None
        
        if self.top == None:
            print("\nStack is empty - nothing to peek")
        else:
             # Get data from the top of the stack
            peeked_data = self.top.get_data()

        return peeked_data


    def output_stack(self):
        """Output the state of the stack"""

        # Testing
        if self.top != None:
            print("------ State of the stack (first item is the top) ------")
            current_node = self.top
            current_node.output_node()
            while current_node.get_next() != None:
                current_node = current_node.get_next()
                current_node.output_node()
            print()
      


        
def main():
    """Create a stack and push, pop and peek items"""
    
    print("### Stack (OOP) ###\n")

    # Initialise the stack
    my_stack = Stack()

    # Insert test data onto the stack
    my_stack.push("Julie")
    my_stack.output_stack()
    
    my_stack.push("Rey")
    my_stack.output_stack()

    # Peek the top item of the stack
    peeked_item = my_stack.peek()
    print(f"Peeked top of the stack: {peeked_item}")
    my_stack.output_stack()
    
    # Insert more test data onto the stack
    my_stack.push("Habib")
    my_stack.output_stack()
    
    my_stack.push("Sabrina")
    my_stack.output_stack()

    # Pop the top item from the stack
    popped_item = my_stack.pop()
    print(f"Popped {popped_item}")
    my_stack.output_stack()
    
    # Insert more test data onto the stack
    my_stack.push("Mina")
    my_stack.output_stack()
    
    my_stack.push("Eirini")
    my_stack.output_stack()

 
# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == '__main__':
    main()
